usr.man/man3/getttyent.3
.\" Copyright (c) 1989, 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by the University of
.\"	California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)getttyent.3	8.1 (Berkeley) 6/4/93
.\"
.TH GETTTYENT 3 "23 February 1996" GNO "Library Routines"
.SH NAME
.BR getttyent ,
.BR getttynam ,
.BR setttyent ,
.BR endttyent
\- get ttys file entry
.SH SYNOPSIS
.br
#include <ttyent.h>
.sp 1
struct ttyent *\fBgetttyent\fR (void);
.br
struct ttyent *\fBgetttynam\fR (char *\fIname\fR);
.br
int \fBsetttyent\fR (void);
.br
int \fBendttyent\fR (void);
.SH DESCRIPTION
The
.BR getttyent ,
and
.BR getttynam 
functions
each return a pointer to an object, with the following structure,
containing the broken-out fields of a line from the tty description
file.
.nf

struct ttyent {
	char	*ty_name;	/* terminal device name */
	char	*ty_getty;	/* command to execute */
	char	*ty_type;	/* terminal type */
#define	TTY_ON		0x01	/* enable logins */
#define	TTY_SECURE	0x02	/* allow uid of 0 to login */
	int	ty_status;	/* flag values */
	char	*ty_window;	/* command for window manager */
	char	*ty_comment;	/* comment field */
};
.fi
.LP
The fields are as follows:
.RS
.IP ty_name
The name of the character-special file.
.IP ty_getty
The name of the command invoked by
.BR init (8)
to initialize tty line characteristics.
.IP ty_type
The name of the default terminal type connected to this tty line.
.IP ty_status
A mask of bit fields which indicate various actions allowed on this
tty line.
The possible flags are as follows:
.RS
.IP TTY_ON
Enables logins (i.e.,
.BR init (8)
will start the command referenced by
.I ty_getty
on this entry).
.IP TTY_SECURE
Allow users with a uid of 0 to login on this terminal.
.RE
.IP ty_window
The command to execute for a window system associated with the line.
.IP ty_comment
Any trailing comment field, with any leading hash marks (``#'') or
whitespace removed.
.RE
.LP
If any of the fields pointing to character strings are unspecified,
they are returned as null pointers.
The field
.I ty_status
will be zero if no flag values are specified.
.LP
See
.BR ttys (5)
for a more complete discussion of the meaning and usage of the
fields.
.LP
The
.BR getttyent 
function
reads the next line from the ttys file, opening the file if necessary.
The
.BR setttyent 
function
rewinds the file if open, or opens the file if it is unopened.
The
.BR endttyent 
function
closes any open files.
.LP
The
.BR getttynam 
function
searches from the beginning of the file until a matching
.I name
is found
(or until
.BR EOF
is encountered).
.SH RETURN VALUES
The routines
.BR getttyent 
and
.BR getttynam 
return a null pointer on
.BR EOF
or error.
The
.BR setttyent 
function
and
.BR endttyent 
return 0 on failure and 1 on success.
.SH FILES
.RS
.BR /etc/ttys
.RE
.SH SEE ALSO
.BR login (1),
.BR ttyslot (3),
.BR gettytab (5),
.BR termcap (5),
.BR ttys (5),
.BR getty (8),
.BR init (8)
.SH HISTORY
The
.BR getttyent ,
.BR getttynam ,
.BR setttyent ,
and
.BR endttyent 
functions appeared in 4.3BSD.
.SH BUGS
These functions use static data storage;
if the data is needed for future use, it should be
copied before any subsequent calls overwrite it.


lib/libc/gen/getttyent.c
/*
 * Copyright (c) 1989 The Regents of the University of California.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by the University of
 *	California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#ifdef __ORCAC__
segment "libc_gen__";
#endif

#if defined(LIBC_SCCS) && !defined(lint)
static char sccsid[] = "@(#)getttyent.c	5.10 (Berkeley) 3/23/91";
#endif /* LIBC_SCCS and not lint */

#define index strchr

#include <ttyent.h>
#include <stdio.h>
#include <ctype.h>
#include <string.h>

static char zapchar;
static FILE *tf;

static char *skip(register char *p);
static char *value(register char *p);

struct ttyent *
getttynam (const char *tty) {
	register struct ttyent *t;

	setttyent();
	while (t = getttyent())
		if (!strcmp(tty, t->ty_name))
			break;
	endttyent();
	return (t);
}

struct ttyent *
getttyent(void) {
	static struct ttyent tty;
	register int c;
	register char *p;
#define	MAXLINELENGTH	100
	static char line[MAXLINELENGTH];

	if (!tf && !setttyent())
		return (NULL);
	for (;;) {
		if (!fgets(p = line, sizeof(line), tf))
			return (NULL);
		/* skip lines that are too big */
		if (!index(p, '\n')) {
			while ((c = getc(tf)) != '\n' && c != EOF)
				;
			continue;
		}
		while (isspace(*p))
			++p;
		if (*p && *p != '#')
			break;
	}
	zapchar = 0;
	tty.ty_name = p;
	p = skip(p);
	if (!*(tty.ty_getty = p))
		tty.ty_getty = tty.ty_type = NULL;
	else {
		p = skip(p);
		if (!*(tty.ty_type = p))
			tty.ty_type = NULL;
		else
			p = skip(p);
	}
	tty.ty_status = 0;
	tty.ty_window = NULL;

#ifdef __ORCAC__
#define	scmp(e)	!strncmp(p, e, strlen(e) ) && isspace(p[strlen(e)])
#define	vcmp(e)	!strncmp(p, e, strlen(e) ) && p[strlen(e)] == '='
#else
#define	scmp(e)	!strncmp(p, e, sizeof e - 1) && isspace(p[sizeof e])
#define	vcmp(e)	!strncmp(p, e, sizeof e - 1) && p[sizeof e] == '='
#endif
	for (; *p; p = skip(p)) {
                if (scmp(_TTYS_OFF))
			tty.ty_status &= ~TTY_ON;
		else if (scmp(_TTYS_ON))
			tty.ty_status |= TTY_ON;
		else if (scmp(_TTYS_SECURE))
			tty.ty_status |= TTY_SECURE;
		else if (vcmp(_TTYS_WINDOW))
			tty.ty_window = value(p);
		else
			break;
	}

	if (zapchar == '#' || *p == '#')
		while ((c = *++p) == ' ' || c == '\t')
			;
	tty.ty_comment = p;
	if (*p == 0)
		tty.ty_comment = 0;
	if (p = index(p, '\r'))
		*p = '\0';
	return (&tty);
}

#define	QUOTED	1

/*
 * Skip over the current field, removing quotes, and return a pointer to
 * the next field.
 */
static char *
skip(register char *p)
{
	register char *t;
	register int c, q;

	for (q = 0, t = p; (c = *p) != '\0'; p++) {
		if (c == '"') {
			q ^= QUOTED;	/* obscure, but nice */
			continue;
		}
		if (q == QUOTED && *p == '\\' && *(p+1) == '"')
			p++;
		*t++ = *p;
		if (q == QUOTED)
			continue;
		if (c == '#') {
			zapchar = c;
			*p = 0;
			break;
		}
		if (c == '\t' || c == ' ' || c == '\n' || c == '\r') {
			zapchar = c;
			*p++ = 0;
			while ((c = *p) == '\t' || c == ' ' || c == '\n' || c == '\r')
				p++;
			break;
		}
	}
	*--t = '\0';
	return (p);
}

static char *
value(register char *p)
{
	return ((p = index(p, '=')) ? ++p : NULL);
}

int
setttyent(void)
{
	if (tf) {
		(void)rewind(tf);
		return (1);
	} else if (tf = fopen(_PATH_TTYS, "r"))
		return (1);
	return (0);
}

int
endttyent(void)
{
int rval;

	if (tf) {
		rval = !(fclose(tf) == EOF);
		tf = NULL;
		return (rval);
	}
	return (1);
}

